File: apps/ios/Mollotov/Handlers/ScrollHandler.swift
import WebKit

/// Handles scroll, scroll2, scrollToTop, scrollToBottom.
struct ScrollHandler {
    let context: HandlerContext

    func register(on router: Router) {
        router.register("scroll") { body in await scroll(body) }
        router.register("scroll2") { body in await scroll2(body) }
        router.register("scroll-to-top") { _ in await scrollTo(top: true) }
        router.register("scroll-to-bottom") { _ in await scrollTo(top: false) }
    }

    @MainActor
    private func scroll(_ body: [String: Any]) async -> [String: Any] {
        let dx = body["deltaX"] as? Double ?? 0
        let dy = body["deltaY"] as? Double ?? 0
        let js = "window.scrollBy(\(dx), \(dy)); ({scrollX: window.scrollX, scrollY: window.scrollY})"
        do {
            let result = try await context.evaluateJSReturningJSON(js)
            return successResponse(result)
        } catch {
            return errorResponse(code: "EVAL_ERROR", message: error.localizedDescription)
        }
    }

    @MainActor
    private func scroll2(_ body: [String: Any]) async -> [String: Any] {
        guard let selector = body["selector"] as? String else {
            return errorResponse(code: "MISSING_PARAM", message: "selector is required")
        }
        let position = body["position"] as? String ?? "center"
        let maxScrolls = body["maxScrolls"] as? Int ?? 10
        let js = """
        (function() {
            var el = document.querySelector('\(selector.replacingOccurrences(of: "'", with: "\\'"))');
            if (!el) return null;
            el.scrollIntoView({block: '\(position)', behavior: 'smooth'});
            var rect = el.getBoundingClientRect();
            return {element: {tag: el.tagName.toLowerCase(), visible: rect.top >= 0 && rect.bottom <= window.innerHeight, rect: {x: rect.x, y: rect.y, width: rect.width, height: rect.height}}, scrollsPerformed: 1, viewport: {width: window.innerWidth, height: window.innerHeight}};
        })()
        """
        do {
            let result = try await context.evaluateJSReturningJSON(js)
            if result.isEmpty { return errorResponse(code: "ELEMENT_NOT_FOUND", message: "Element not found: \(selector)") }
            return successResponse(result)
        } catch {
            return errorResponse(code: "EVAL_ERROR", message: error.localizedDescription)
        }
    }

    @MainActor
    private func scrollTo(top: Bool) async -> [String: Any] {
        let js = top
            ? "window.scrollTo(0, 0); ({scrollY: 0})"
            : "window.scrollTo(0, document.documentElement.scrollHeight); ({scrollY: window.scrollY})"
        do {
            let result = try await context.evaluateJSReturningJSON(js)
            return successResponse(result)
        } catch {
            return errorResponse(code: "EVAL_ERROR", message: error.localizedDescription)
        }
    }
}

File: native/engine-chromium-desktop/src/handlers/scroll_handler.cpp
#include "scroll_handler.h"

namespace mollotov {

ScrollHandler::ScrollHandler(DesktopHandlerRuntime runtime) : runtime_(std::move(runtime)) {}

void ScrollHandler::Register(DesktopRouter& router) const {
  router.Register("scroll", [this](const nlohmann::json& params) { return Scroll(params); });
  router.Register("scroll-to-top",
                  [this](const nlohmann::json&) { return ScrollTo("window.scrollTo(0, 0)"); });
  router.Register("scroll-to-bottom",
                  [this](const nlohmann::json&) {
                    return ScrollTo("window.scrollTo(0, document.body.scrollHeight)");
                  });
  router.Register("scroll2", [](const nlohmann::json&) { return Unsupported("scroll2"); });
}

nlohmann::json ScrollHandler::Scroll(const nlohmann::json& params) const {
  HandlerContext& context = RequireHandlerContext(runtime_);
  const int delta_x = IntOrDefault(params, "deltaX", 0);
  const int delta_y = IntOrDefault(params, "deltaY", 0);
  const std::string script =
      "(() => {"
      "window.scrollBy(" + std::to_string(delta_x) + ", " + std::to_string(delta_y) + ");"
      "return {scrollX: window.scrollX, scrollY: window.scrollY};"
      "})()";
  const nlohmann::json result = context.EvaluateJsReturningJson(script);
  return SuccessResponse({
      {"scrollX", result.value("scrollX", delta_x)},
      {"scrollY", result.value("scrollY", delta_y)},
  });
}

nlohmann::json ScrollHandler::ScrollTo(const std::string& expression) const {
  HandlerContext& context = RequireHandlerContext(runtime_);
  const std::string script =
      "(() => {" + expression + "; return {scrollY: window.scrollY}; })()";
  const nlohmann::json result = context.EvaluateJsReturningJson(script);
  return SuccessResponse({{"scrollY", result.value("scrollY", 0)}});
}

}  // namespace mollotov
